ehcache reconnect changes

mobasherul-ibm opened the pull request.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/CacheStateRepositoryReconnectTest.java:
```diff
@@ -0,0 +1,196 @@
+
+  @ClassRule
+  @Rule
+  public static final TestRetryer<Duration, Cluster> CLUSTER = tryValues(ofSeconds(5)).map(leaseLength -> newCluster().in(clusterPath()).withServiceFragment(offheapResource("primary-server-resource", 64) + leaseLength(leaseLength)).build()).outputIs(CLASS_RULE);
```
This line is more than 260 characters long -- it should be split apart so the lines are between 80 and 120 characters long.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/CacheStateRepositoryReconnectTest.java:
```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright Terracotta, Inc.
+ * Copyright IBM Corp. 2024, 2025
```
As a new class, should this simply be 2025?

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/CacheStateRepositoryReconnectTest.java:
```diff
@@ -0,0 +1,196 @@
+
+    assertThat(cache, notNullValue());
+
+    cacheManager.destroyCache("clustered-cache");
```
This should be in a `finally` block ...

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/CacheStateRepositoryReconnectTest.java:
```diff
@@ -0,0 +1,196 @@
+
+    CacheManagerBuilder<PersistentCacheManager> clusteredCacheManagerBuilder = CacheManagerBuilder.newCacheManagerBuilder().with(ClusteringServiceConfigurationBuilder.cluster(connectionURI.resolve("/crud-cm")).autoCreate(server -> server.defaultServerResource("primary-server-resource")));
+    cacheManager = clusteredCacheManagerBuilder.build(false);
+    cacheManager.init();
```
I believe there should be an `AfterClass` that closes this cache manager.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/ThrowingResiliencyStrategy.java:
```diff
@@ -39,7 +39,7 @@ public boolean containsKeyFailure(K key, StoreAccessException e) {
   @Override
   public void putFailure(K key, V value, StoreAccessException e) {
-    if (e.getCause() instanceof ReconnectInProgressException) {
+    if (e.getCause().getCause() instanceof ReconnectInProgressException || e.getCause() instanceof ReconnectInProgressException) {
```
The additional level of exception testing bothers me.  Can it be avoided?

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/IterationFailureBehaviorTest.java:
```diff
@@ -120,8 +120,8 @@ public void testIteratorFailover() throws Exception {
-        assertThat(e.getCause().getCause().getCause().getCause(),
-          either(instanceOf(ConnectionClosedException.class)) //lost in the space between active and passive
+        assertThat(e.getCause().getCause().getCause(),
+          either(instanceOf(ReconnectInProgressException.class)) //lost in the space between active and passive
```
Can you explain the need for this change?

cljohnso replied:
So does `ConnectionClosedException` no longer appear in the exception cause chain?  If not, maybe it should.

cljohnso replied:
@chrisdennis Is there any reason for the `ReconnectInProgressException` used by `ReconnectInProgressClusterTierClientEntity` include the exception causing the reconnect attempt, i.e. the `ConnectionClosedException`?

mobasherul-ibm commented:
> The changes to the exception handling here seem like they might be amplifying and obfuscating these changes. If we can manage to preserve much of the original exception throwing/logic I feel we'll have a smaller more trustable and reviewable changeset.

 Earlier flow was store -> reconnectingProxy -> commonProxy -> actualentity where exception flow was -
(StoreAccessException <- ReconnectInProgressException) when connection breaks Or for other scenarios 
(StoreAccessException<-ServerStoreProxyException<-SomeOtherException)  because commonProxy wraps ConnectionClosedException with ServerStoreProxyException and that is passed to reconnectingProxy which just throws ReconnectInProgressException.

Since I removed reconnectingProxy and introduced reconnectable entity flow has changed to - 
store -> commonProxy -> reconnectableentity -> actual entity where exception flow has changed to - 
(StoreAccessException <- ServerStoreProxyException <- ReconnectInProgressException) when connection breaks as commonProxy is now wrapping the ReconnectInProgressException with ServerStoreProxyException and for other exception it will be -
(StoreAccessException <-. ServerStoreProxyException <- SomeOtherException).

mobasherul-ibm commented on clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/reconnect/FailedReconnectClusterTierClientEntity.java:
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Terracotta, Inc.
+ * Copyright IBM Corp. 2024, 2025
```
that fails copyright checks

mobasherul-ibm commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/BasicCacheReconnectTest.java:
```diff
@@ -125,7 +126,7 @@ public void cacheOpsDuringReconnection() throws Exception {
       });
 
-      getSucceededFuture.get(20000, TimeUnit.MILLISECONDS);
+      assertThat(getSucceededFuture::isDone, eventually().is(true));
```
didn't get that exception will be thrown for some time since reconnection is happening in  background. I changed this since it can be slow at times for reconnect to complete. Eventually will still throw the timeout exception as it waits for 1 min for the future to succeed.

mobasherul-ibm commented on clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/service/ConnectionState.java:
```diff
@@ -265,8 +268,9 @@ public void destroyState(boolean healthyConnection) {
     entityFactory = null;
-
-    clusterTierEntities.clear();
+    if (healthyConnection) {
```
because healthy is passed as false for reconnect and we only want to swap the new actual entity inside the reconnectable  entities.

mobasherul-ibm commented on clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/reconnect/FailedReconnectClusterTierClientEntity.java:
```diff
@@ -0,0 +1,107 @@
+  private final String cacheId;
+  private final Throwable failure;
+
+  public FailedReconnectClusterTierClientEntity(String cacheId, Throwable failure) {
```
this i just tried to keep it similar with earlier FailedReconnectStoreProxy can change it yes.

mobasherul-ibm commented on clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/reconnect/FailedReconnectClusterTierClientEntity.java:
```diff
@@ -0,0 +1,107 @@
+
+  @Override
+  public <T extends EhcacheEntityResponse> void addResponseListener(Class<T> responseType, ResponseListener<T> responseListener) {
+    throw new RuntimeException("Cache " + getCacheId() + " failed reconnecting to cluster", failure);
```
I guess its fine any way these references are unusable now.

cljohnso commented on clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/reconnect/ReconnectableClusterTierClientEntity.java:
```diff
@@ -0,0 +1,276 @@
+
+    @Override
+    public void close() {
+      throw new ReconnectInProgressException();
```
Is throwing the right thing to do here?  If an application wishes to close the entity while a reconnect is in progress, isn't a reasonable action to cancel the reconnect attempt and actually complete the close?

cljohnso commented:
@mobasherul-ibm please do not resolve comments you did not create -- it makes re-review considerably more difficult.  If you want to keep track of items you've addressed, reply to the comment with a `+1` or a `:white_check_mark`.

cljohnso commented:
Still have a test failure in `org.ehcache.clustered.IterationFailureBehaviorTest#testIteratorReconnect`

mobasherul-ibm commented on clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/reconnect/ReconnectableClusterTierClientEntity.java:
```diff
@@ -0,0 +1,276 @@
+    } catch (Exception sspe) {
+      if (containsCause(sspe, ConnectionClosedException.class)) {
+        if (delegateRef.compareAndSet(cl, new ReconnectInProgressClusterTierClientEntity())) {
+          CompletableFuture.runAsync(reconnectTask, asyncExecutor);
```
Single thread executor ensures reconnect via two caches can't happen in parallel

mobasherul-ibm replied:
This executor is used for one more task i.e executing the responses like -cache invalidation etc. received from server on cache entities fetched from same connection .

mobasherul-ibm commented on clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/service/DefaultClusteringService.java:
```diff
@@ -282,13 +285,45 @@ public <K, V> ServerStoreProxy getServerStoreProxy(ClusteredCacheIdentifier cach
+        //TODO: handle race between disconnect event and connection closed exception being thrown
+        // this guy should wait till disconnect event processing is complete.
+        ReconnectableClusterTierClientEntity reconnectableEntity = (ReconnectableClusterTierClientEntity) connectionState.getClusterTierClientEntity(cacheId);
+        synchronized (reconnectableEntity) {
```
removing it

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java:
```diff
@@ -0,0 +1,199 @@
+        cache2.put(0L, "secondcache");
+      });
+
+      try {
```
This form would be preferable to the try/catch block:

```java
      assertThat(() -> putFuture1.get(5000, TimeUnit.MILLISECONDS),
        threw(Matchers.<Throwable>both(instanceOf(ExecutionException.class))
          .and(hasCause(hasCause(hasCause(hasCause(instanceOf(ReconnectInProgressException.class))))))));
```

@chrisdennis recommend a different mechanism than the chain of `hasCause` methods.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java:
```diff
@@ -0,0 +1,199 @@
+        assertThat(e.getCause().getCause().getCause().getCause(), instanceOf(ReconnectInProgressException.class));
+      }
+
+      try {
```
See comment for section above.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java:
```diff
@@ -0,0 +1,199 @@
+      cache2.put(0L, "secondValue");
+      expireLease();
+
+      try {
```
`assertThat(..., threw(...))` would be preferred here.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java:
```diff
@@ -0,0 +1,199 @@
+
+      cacheManager.removeCache("clustered-cache1");
+
+      try {
```
See comment on block above.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/replication/BasicClusteredCacheOpsReplicationTest.java:
```diff
@@ -190,6 +191,7 @@ public void testCAS() throws Exception {
   }
 
   @Test
+  @Ignore
```
Is this being ignored with expectation for a future repair?  If so, an issue number should be mentioned.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/replication/BasicClusteredCacheOpsReplicationMultiThreadedTest.java:
```diff
@@ -241,6 +242,7 @@ public void testBulkOps() throws Exception {
   }
 
   @Test
+  @Ignore
```
Is this being ignored with expectation for a future repair?  If so, an issue number should be mentioned.

cljohnso commented on clustered/integration-test/build.gradle:
```diff
@@ -69,7 +69,7 @@ test {
   systemProperty 'kitInstallationPath', "$unzipKit.destinationDir/${project(':clustered:ehcache-clustered').archivesBaseName}-$project.version-kit"
   // Uncomment to include client logging in console output
-  // testLogging.showStandardStreams = true
+   testLogging.showStandardStreams = true
```
Permanent change?

mobasherul-ibm replied:
will remove once approved

mobasherul-ibm commented on clustered/integration-test/src/test/java/org/ehcache/clustered/IterationFailureBehaviorTest.java:
```diff
@@ -80,7 +80,7 @@ public void testIteratorFailover() throws Exception {
       = CacheManagerBuilder.newCacheManagerBuilder()
       .with(ClusteringServiceConfigurationBuilder.cluster(CLUSTER.get().getConnectionURI().resolve("/iterator-cm"))
-        .autoCreate(server -> server.defaultServerResource("primary-server-resource"))
+        .autoCreateOnReconnect(server -> server.defaultServerResource("primary-server-resource"))
```
@cljohnso I did this in the test because test is killing all the servers and then starting them .

mobasherul-ibm replied:
yeah autoCreateOnReconnect is not needed in this one.

mobasherul-ibm commented on clustered/integration-test/src/test/java/org/ehcache/clustered/IterationFailureBehaviorTest.java:
```diff
@@ -137,7 +137,7 @@ public void testIteratorReconnect() throws Exception {
       = CacheManagerBuilder.newCacheManagerBuilder()
       .with(ClusteringServiceConfigurationBuilder.cluster(CLUSTER.get().getConnectionURI().resolve("/iterator-cm"))
-        .autoCreate(server -> server.defaultServerResource("primary-server-resource")));
+        .autoCreateOnReconnect(server -> server.defaultServerResource("primary-server-resource")));
```
earlier entity were not reconnectable so there were no connection retries but now that changed thats why I changed it to ensure all connections are closed at the end of this test.

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java:
```diff
@@ -0,0 +1,206 @@
+        cache2.put(0L, "secondcache");
+      });
+
+      assertThat(() -> putFuture1.get(5000, TimeUnit.MILLISECONDS),
```
Is this going to need an `eventually` to have a stable test?

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java:
```diff
@@ -0,0 +1,206 @@
+        threw(Matchers.<Throwable>both(instanceOf(ExecutionException.class))
+          .and(hasCause(hasCause(hasCause(hasCause(instanceOf(ReconnectInProgressException.class))))))));
+
+      assertThat(() -> putFuture2.get(5000, TimeUnit.MILLISECONDS),
```
Is this going to need an `eventually` to have a stable test?

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java:
```diff
@@ -0,0 +1,206 @@
+      cache2.put(0L, "secondValue");
+      expireLease();
+
+      assertThat(() -> cache1.get(0L),
```
Is it possible the reconnect succeeds  before this statement runs?

cljohnso commented on clustered/integration-test/src/test/java/org/ehcache/clustered/reconnect/MultipleCacheReconnectTest.java:
```diff
@@ -0,0 +1,206 @@
+
+      cacheManager.removeCache("clustered-cache1");
+
+      try {
```
This would be better with the `assertThat(..., threw(...))` pattern.

cljohnso reviewed (commented):
Needs conversations with Chris and possibly some minor tweaks.
